python/paddle/v2/fluid/layers/math_op_patch.py

```diff
@@ -0,0 +1,161 @@
+            dtype = var.dtype
+        except:
+            raise ValueError("Cannot get data type from %s", var.name)
+        return dtype
+
+    def create_scalar(block, value, dtype):
+        value = float(value)
+        tmp_name = new_name()
+        var = block.create_var(name=tmp_name, shape=[1], dtype=dtype)
+        block.append_op(
+            type="fill",
+            outputs={"Out": [var]},
+            attrs={"value": [value],
+                   "shape": [1],
+                   "dtype": dtype})
+        return var
+
+    def create_tensor(block, value, dtype, shape):
+        value = float(value)
+        tmp_name = new_name()
+        var = block.create_var(name=tmp_name, shape=shape, dtype=dtype)
+        block.append_op(
+            type="fill_constant",
+            outputs={'Out': [var]},
+            attrs={'dtype': var.dtype,
+                   'shape': shape,
+                   'value': value})
+        return var
+
+    def create_tensor_with_batchsize(ref_var, value, dtype):
+        assert isinstance(ref_var, Variable)
+        value = float(value)
+        tmp_name = new_name()
+        var = ref_var.block.create_var(name=tmp_name, dtype=dtype)
+        ref_var.block.append_op(
+            type='fill_constant_batch_size_like',
+            outputs={'Out': [var]},
+            inputs={'Input': [ref_var]},
+            attrs={'shape': ref_var.shape,
+                   'value': value})
+        return var
+
+    def astype(self, dtype):
+        """
+        Cast a variable to data type.
+        NOTE: The variable must be a Tensor
+        Args:
+            self(Variable): The source variable
+            dtype: The target dtype
+
+        Returns:
+            Variable with new dtype
+        """
+        tmp_name = new_name()
+        out = self.block.create_var(name=tmp_name, dtype=dtype)
+        self.block.append_op(
+            type="cast",
+            inputs={"X": [self]},
+            outputs={"Out": [out]},
+            attrs={"in_dtype": self.dtype,
+                   "out_dtype": out.dtype})
+        return out
+
+    def _elemwise_method_creator_(method_name, op_type, reverse=False):
+        def __impl__(self, other_var):
+            lhs_dtype = safe_get_dtype(self)
+
+            if not isinstance(other_var, Variable):
+                if reverse:
+                    has_batch_size = False
+                    for elem in self.shape:
+                        if elem < 0:
+                            has_batch_size = True
+                            break
+                    if not has_batch_size:
+                        other_var = create_tensor(
+                            self.block,
+                            other_var,
+                            dtype=lhs_dtype,
+                            shape=self.shape)
+                    else:
+                        other_var = create_tensor_with_batchsize(
+                            self, other_var, lhs_dtype)
+                else:
```

Comment:
Why does the `other_var` is cast to `Variable` by different methods in the two branches of `if reverse:`?

Reply:
The reversing operator in Python means the left-hand operand is a Python value and the right-hand operand is a Variable.

In the elementwise-operators, the right-hand operand can be broadcasted. Supposed we write an `a - 10`, every element in `a` will be minus 10.  However, for `10 - a`, we do not implement the broadcast in C++ operators. So we should explicitly broadcast `10` to the same shape of `a`.

Getting the shape of a variable has two situations. 

1. The variable is a parameter. Its shape is decided at compile time and not related to the batch size. So we use `fill_constant`.
2. The variable is a layer output. Its shape is decided at runtime and related to the batch size. So we use `fill_constant_batch_size_like`.

Reply:
Great idea. Impressive.